_source/_posts/2022-09-21-angular-dependency-injection.md

```diff
@@ -0,0 +1,296 @@
+
+```ts
+@Injectable({
+  providedIn: 'root'
+})
+export class AuthService {
+  public login(): void { }
+  public logout(): void { }
+}
+```
+
+In a stroke of luck, a large tech company decides to buy your language learning app, but they require you to authenticate using their social login only. You can create a new authentication service; we'll call it `NewAuthService` (note, you should not name your services with these terrible names). 
+
+```ts
+@Injectable({
+  providedIn: 'root'
+})
+export class NewAuthService {
+  public login(): void { /* new way to login */ }
+  public logout(): void { /* new way to logout */ }
+```
+
+Because both classes have the same public members, you can substitute the original `AuthService` with the new `NewAuthService` by configuring the provider:
+
+```ts
+@NgModule({
+  imports: // imports here
+  declarations: //declarations here
+  providers: [
+    { provide: AuthService, useClass: NewAuthService }
+  ]
+})
+export class AppModule { }
+```
+
+The cool thing about having the same public members is that there's no need to change the consuming code. Angular instantiates a new instance of `NewAuthService` and provides that dependency to consuming code, even if they still refer to `AuthService`!
+
+### Configure providers with `useExisting`
+
+The `useExisting` option replaces the provider with a different provider already existing within the application.  This option is a great use case for API narrowing, that is, decreasing the surface area of an API. Let's say your language learning application has an API that got out of hand. We'll call this API `LanguageTranslationService`, and it looks like this:
+
+```ts
+@Injectable({
+  providedIn: 'root'
+})
+export const LanguageTranslationService {
+  public french(text: string): string { /* translates to French */ }
+  public japanese(text: string): string { /* translates to Japanese */ }
+  public elvish(text: string): string { /* translates to Elvish */ }
+  public klingon(text: string): string { /* translates to Klingon */ }
+  // so on and so forth
+}
+```
+
+And you consume the service like this:
+
+```ts
+@Component({
+  // standard component metadata here
+})
+export class ElvishTranslationComponent implements OnInit {
+  private elvish!: string;
+  constructor(private translationService: LanguageTranslationService) { }
+  
+  public ngOnInit(): void {
+    this.elvish = this.translationService.elvish(someText);
+  }
+}
+```
+
+Oops... The `LanguageTranslationService` looks a bit unwieldy. Let's narrow the API surface by creating a new class called `FictiousLanguageTranslationService` and move the translation methods for the fictitious languages there. We'll use an abstract class for this:
```

Comment:
```suggestion
Oops... The `LanguageTranslationService` looks a bit unwieldy. Let's narrow the API surface by creating a new class called `FictitiousLanguageTranslationService` and move the translation methods for the fictitious languages there. We'll use an abstract class for this:
```
since fictious is a fictitious word.... (sorry!) i'm going to edit the class name in the code below also....